crates/types/src/schema/metadata/mod.rs

```diff
@@ -0,0 +1,566 @@
+use crate::schema::invocation_target::{
+    DEFAULT_IDEMPOTENCY_RETENTION, DEFAULT_WORKFLOW_COMPLETION_RETENTION, InputRules,
+    InvocationTargetMetadata, InvocationTargetResolver, OutputRules,
+};
+use crate::schema::metadata::openapi::ServiceOpenAPI;
+use crate::schema::service::ServiceMetadataResolver;
+use crate::schema::subscriptions::{ListSubscriptionFilter, Subscription, SubscriptionResolver};
+use crate::schema::{deployment, service};
+use crate::time::MillisSinceEpoch;
+use crate::{Version, Versioned, identifiers};
+use arc_swap::ArcSwapOption;
+use restate_serde_util::MapAsVecItem;
+use serde_json::Value;
+use serde_with::serde_as;
+use std::collections::HashMap;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+use std::time::Duration;
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(from = "serde_hacks::Schema", into = "serde_hacks::Schema")]
+pub struct Schema {
+    /// This gets bumped on each update.
+    version: Version,
+
+    deployments: HashMap<DeploymentId, Deployment>,
+    active_service_revisions: HashMap<String, ActiveServiceRevision>,
+    subscriptions: HashMap<SubscriptionId, Subscription>,
+}
+
+impl Default for Schema {
+    fn default() -> Self {
+        Self {
+            version: Version::INVALID,
+            active_service_revisions: HashMap::default(),
+            deployments: HashMap::default(),
+            subscriptions: HashMap::default(),
+        }
+    }
+}
+
+impl GlobalMetadata for Schema {
+    const KEY: &'static str = "schema_registry";
+
+    const KIND: MetadataKind = MetadataKind::Schema;
+
+    fn into_container(self: Arc<Self>) -> MetadataContainer {
+        MetadataContainer::Schema(self)
+    }
+}
+
+impl Versioned for Schema {
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+mod storage {
+    use crate::flexbuffers_storage_encode_decode;
+
+    use super::Schema;
+
+    flexbuffers_storage_encode_decode!(Schema);
+}
+
+// -- Data model
+
+#[derive(Debug, Clone)]
+struct ActiveServiceRevision {
+    deployment_id: DeploymentId,
+    service_revision: Arc<ServiceRevision>,
+}
+
+impl ActiveServiceRevision {
+    fn as_service_metadata(&self) -> service::ServiceMetadata {
+        self.service_revision
+            .as_service_metadata(self.deployment_id)
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct Deployment {
+    id: DeploymentId,
+    ty: DeploymentType,
+    delivery_options: DeliveryOptions,
+    supported_protocol_versions: RangeInclusive<i32>,
+    /// Declared SDK during discovery
+    sdk_version: Option<String>,
+    created_at: MillisSinceEpoch,
+
+    #[serde_as(as = "restate_serde_util::MapAsVec")]
+    services: HashMap<String, Arc<ServiceRevision>>,
+}
+
+impl MapAsVecItem for Deployment {
+    type Key = DeploymentId;
+
+    fn key(&self) -> Self::Key {
+        self.id
+    }
+}
+
+impl Deployment {
+    fn as_deployment(&self) -> deployment::Deployment {
```

Comment:
For the uninitiated; what does `Deployment::as_deployment()` mean? additionally, `as_*` conventionally is used for [zero-cost conversions.](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)